[WPE] wpe_renderer_backend_egl_target_deinitialize() is never called

aperezdc opened the pull request:
#### ff2fcc374df4847f082284ab22f2925d8f5b2ef1
<pre>
[WPE] wpe_renderer_backend_egl_target_deinitialize() is never called
<a href="https://bugs.webkit.org/show_bug.cgi?id=298225">https://bugs.webkit.org/show_bug.cgi?id=298225</a>

Reviewed by Carlos Garcia Campos.

* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
(WebKit::AcceleratedSurface::RenderTargetWPEBackend::~RenderTargetWPEBackend):
Add missing call to wpe_renderer_backend_egl_target_deinitialize().

Canonical link: <a href="https://commits.webkit.org/299801@main">https://commits.webkit.org/299801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c6820924b37394419ce33be43074d35e8c71475

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30577 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126590 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72305 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48503 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/91306 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60604 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107793 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/71860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31475 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70215 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26084 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129486 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47152 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35775 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99927 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103978 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99769 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25329 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45222 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23249 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43783 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47015 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/52720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46482 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49829 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (commented):
Uhh... this seems weird. Probably WPERendererBackend ought to take care of this itself if it hasn't been "deinitialized" before destroy is called. At least that's the pattern we would use in any GObject class, or any WebKit C++ class.

mcatanzaro commented:
Possibly.

This looks like a "pre destroy" pattern: where you provide a way to optionally deinitialize an object whenever, separate from destruction. Right? It's conventional for destroying such an object to also deinitialize it, such that a deinitialize called immediately before destruction is redundant. Because why not? It would be silly to ever allow the object to possibly enter a partially-destroyed state. So yes, if adding the deinitialize call here is necessary, then that seems like a design flaw to me.

There are two ways to do it: (a) every backend deinitializes at the top of destroy, which is what I would do if designing this today, or (b) `wpe_renderer_backend_egl` calls the deinitialize vfunc before calling the destroy vfunc. Either way works... but both ways require auditing every backend to make sure they're behaving as expected.

aperezdc commented:
> Possibly.
> 
> This looks like a "pre destroy" pattern: where you provide a way to optionally deinitialize an object whenever, separate from destruction. Right? It's conventional for destroying such an object to also deinitialize it, such that a deinitialize called immediately before destruction is redundant. Because why not? It would be silly to ever allow the object to possibly enter a partially-destroyed state. So yes, if adding the deinitialize call here is necessary, then that seems like a design flaw to me.
> 
> There are two ways to do it: (a) every backend deinitializes at the top of destroy, which is what I would do if designing this today, or (b) `wpe_renderer_backend_egl` calls the deinitialize vfunc before calling the destroy vfunc. Either way works... but both ways require auditing every backend to make sure they're behaving as expected.

Why would the call to `_deinitialize()` be done inside libwpe? That doesn't make sense, the function is intended to be called *by WebKit*, so if we make libwpe call it and then at some point we wanted to actually call `_deinitialize()` elsewhere in WebKit, we would need to change libwpe to prevent the `::deinitialize` callback being invoked twice.

I hope this helps clarify.

aperezdc commented:
> It does not.
> 
> I would leave this alone. If deinitialize then destroy is not equivalent to destroy, surely the backend should be fixed. Requiring WebKit to deinit immediately before destroy is fragile and strange.

Why? You're not explaining your point of view.

Meanwhile, I have a good reason to add the call, which is fulfilling the API contract in which WebKit is supposed to tell the renderer target when it's a good moment to deiniatilize&mdash;which may or not be immediately before destruction. Currently it is, but we are free to change that if needed later on.

aperezdc commented:
> I'm afraid I am completely bamboozled, Adrian.
> 
> I have no clue why the renderer target needs to be told that now is a good moment to deinitialize if it's about to be destroyed one line down. This just doesn't make any sense to me at all. And how can it be an API contract when this library was created for use by WebKit and WebKit has never done that before this pull request...?
> 
> (I don't actually care too much about this. Feel free to land if you want. It just seems _weird._)

We had a prototype of DMA-BUF usage in WPEBackend-fdo which relied on this, but the version that got merged took a different approach. Yet the API was added in `libwpe`, which means we should be calling `_deainitlize()`. Then some WPE backends started using it, assuming that the callback would be called...

Bit of a pity that we got the `libwpe` merged and in releases if it wasn't strictly needed; but the cat is out of the bag, so we should fulfill the expectation of backend implementors that the `::deinitialize` callback gets triggered. I am going to double check that WPEBackend-fdo at least behaves well with the change, and merge this.

aperezdc committed:
[WPE] wpe_renderer_backend_egl_target_deinitialize() is never called
https://bugs.webkit.org/show_bug.cgi?id=298225

Reviewed by Carlos Garcia Campos.

* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
(WebKit::AcceleratedSurface::RenderTargetWPEBackend::~RenderTargetWPEBackend):
Add missing call to wpe_renderer_backend_egl_target_deinitialize().

Canonical link: https://commits.webkit.org/299801@main
